src/coreclr/jit/morph.cpp

```diff
@@ -9273,6 +9296,127 @@ GenTree* Compiler::fgMorphCall(GenTreeCall* call)
     return call;
 }
 
+/*****************************************************************************
+ *
+ *  Expand and return the call target address for a VirtualCall
+ *  The code here should match that generated by LowerVirtualVtableCall
+ */
+
+GenTree* Compiler::fgExpandVirtualVtableCallTarget(GenTreeCall* call)
+{
+    GenTree* result;
+
+    JITDUMP("Expanding virtual call target for %d.%s:\n", call->gtTreeID, GenTree::OpName(call->gtOper));
+
+    noway_assert(call->gtCallType == CT_USER_FUNC);
+
+// get a reference to the thisPtr being passed
+#if 0
+    GenTree* thisPtr = call->gtCallThisArg->GetNode();
+#else
+    fgArgTabEntry* thisArgTabEntry = gtArgEntryByArgNum(call, 0);
+    GenTree*       thisPtr         = thisArgTabEntry->GetNode();
+#endif
+
+    // Make a copy of the thisPtr by cloning
+
+    assert(thisPtr->OperIsLocal()); // fgMorphArgs must enforce this invariant
+    thisPtr = gtClone(thisPtr, true);
+
+    noway_assert(thisPtr != nullptr);
+
+    // Get hold of the vtable offset (note: this might be expensive)
+    unsigned vtabOffsOfIndirection;
+    unsigned vtabOffsAfterIndirection;
+    bool     isRelative;
+    info.compCompHnd->getMethodVTableOffset(call->gtCallMethHnd, &vtabOffsOfIndirection, &vtabOffsAfterIndirection,
+                                            &isRelative);
+
+    // Dereference the this pointer to obtain the method table, it is called vtab below
+    GenTree* vtab;
+    assert(VPTR_OFFS == 0); // We have to add this value to the thisPtr to get the methodTable
+    vtab = gtNewOperNode(GT_IND, TYP_I_IMPL, thisPtr);
+    vtab->gtFlags |= GTF_IND_NONFAULTING;
+    vtab->gtFlags |= GTF_IND_INVARIANT;
+
+    // Get the appropriate vtable chunk
+    if (vtabOffsOfIndirection != CORINFO_VIRTUALCALL_NO_CHUNK)
+    {
+        if (isRelative)
+        {
+            // MethodTable offset is a relative pointer.
+            //
+            // Additional temporary variable is used to store virtual table pointer.
+            // Address of method is obtained by the next computations:
+            //
+            // Save relative offset to tmp (vtab is virtual table pointer, vtabOffsOfIndirection is offset of
+            // vtable-1st-level-indirection):
+            // tmp = vtab
+            //
+            // Save address of method to result (vtabOffsAfterIndirection is offset of vtable-2nd-level-indirection):
+            // result = [tmp + vtabOffsOfIndirection + vtabOffsAfterIndirection + [tmp + vtabOffsOfIndirection]]
+            //
+            //
+            // If relative pointers are also in second level indirection, additional temporary is used:
```

Comment:
I realize this just mirrors what is over in `LowerVirtualVtableCall`, but I don't undertstand the `If` here -- there is no conditional code that follows.

Reply:
I changed the comment to

`  // When isRelative is true we need to setup two temporary variables `

Reply:
The are two cases being covered here the 'isRealtive' case and the normal case.
Currently I have never seen 'isRelative' being true.  I will add an assert and look into when it is used by the VM.

Reply:
Not sure if we ever test these paths.

cc @jkotas

Reply:
These paths were added by Samsung. There were only ever used by Tizen armel target.

Reply:
@jkotas  Should I just replace the 'isRelative' code path with a `noway_assert` and a comment?

Reply:
I think we should either keep it working or delete it from everywhere if it is no longer used.

@alpencolt Are you still using the relative pointers?

Reply:
In our **dotnet/runtime** source code, I don't see any code that ever sets 'isRelative' to true.
Is their code (**Samsung**) that uses this code path this checked in?

Reply:
It is set here: https://github.com/dotnet/runtime/blob/bf9f899a998eba28f84c77e0c4996d1c22d94b0f/src/coreclr/vm/jitinterface.cpp#L8830

Reply:
> @alpencolt Are you still using the relative pointers?

We use it in FNV (CoreCLR 3.1), but this mode doesn't work in .NET 6 now.
If we'll improve R2R performance relative pointers will not be used. Otherwise we'll have to fix FNV.

Reply:
@jkotas Yes, But I wasn't able to see how `MethodTable::VTableIndir_t::isRelative` was ever set to non-zero

Reply:
It is set to non-zero when `FEATURE_NGEN_RELOCS_OPTIMIZATIONS` is defined.

Reply:
@jkotas  Maybe it gets set in some tricky way that I couldn't figure out. 
 Anyway I will leave support in for this until we get a better understanding of what Samsung's plans are for this.